File: 1301-1400/1367-Linked List in Binary Tree/1367-Linked List in Binary Tree.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def isSubPath(self, head: ListNode, root: TreeNode) -> bool:
        if root is None:
            return False
        def dfs(head, root):
            if head is None:
                return True
            if root is None:
                return False
            if head.val != root.val:
                return False
            return dfs(head.next, root.left) or dfs(head.next, root.right)
        return dfs(head, root) or self.isSubPath(head, root.left) or self.isSubPath(head, root.right)


class Solution2:
    def isSubPath(self, head: ListNode, root: TreeNode) -> bool:
        lps = [0]
        vals = [head.val]
        l = 0
        node = head.next
        while node:
            while l and node.val != vals[l]:
                l = lps[l - 1]
            l += head.val == vals[l]
            lps.append(l)
            vals.append(node.val)            
            node = node.next

        def dfs(root, l):
            if root is None:
                return False
            while l and root.val != vals[l]:
                l = lps[l - 1]
            l += root.val == vals[l]
            if l == len(vals):
                return True
            return dfs(root.left, l) or dfs(root.right, l)
        return dfs(root, 0)

File: 1301-1400/1367-Linked List in Binary Tree/1367-Linked List in Binary Tree.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSubPath(ListNode* head, TreeNode* root) {
        vector<int> lps(1), vals(1, head->val);
        int l = 0;
        ListNode* node = head->next;
        while (node) {
            while (l && node->val != vals[l]) {
                l = lps[l - 1];
            }
            l += node->val == vals[l];
            lps.push_back(l);
            vals.push_back(node->val);
            node = node->next;
        }
        
        return dfs(root, 0, lps, vals);
    }

private:
    bool dfs(TreeNode* root, int l, vector<int>& lps, vector<int>& vals) {
        if (root == nullptr) {
            return false;
        }
        while (l && root->val != vals[l]) {
            l = lps[l - 1];
        }
        l += root->val == vals[l];
        if (l == vals.size()) {
            return true;
        }
        return dfs(root->left, l, lps, vals) || dfs(root->right, l, lps, vals);
    }
};


class Solution2 {
public:
    bool isSubPath(ListNode* head, TreeNode* root) {
        vector<int> lps(1), vals(1, head->val);
        int l = 0;
        ListNode* node = head->next;
        while (node) {
            while (l && node->val != vals[l]) {
                l = lps[l - 1];
            }
            l += node->val == vals[l];
            lps.push_back(l);
            vals.push_back(node->val);
            node = node->next;
        }
        
        return dfs(root, 0, lps, vals);
    }

private:
    bool dfs(TreeNode* root, int l, vector<int>& lps, vector<int>& vals) {
        if (root == nullptr) {
            return false;
        }
        while (l && root->val != vals[l]) {
            l = lps[l - 1];
        }
        l += root->val == vals[l];
        if (l == vals.size()) {
            return true;
        }
        return dfs(root->left, l, lps, vals) || dfs(root->right, l, lps, vals);
    }
};
